### ISSUE TITLE
Declaring a BigQueryDataset without a spec.access field freezes the existing spec.access contents in place

### ISSUE THREAD

**Issue by dflemstr** [GoogleCloudPlatform/k8s-config-connector#465]
## Checklist

- [x] I did not find a related open issue.
- [x] I did not find a solution in the troubleshooting guide: (https://cloud.google.com/config-connector/docs/troubleshooting)
- [x] If this issue is time-sensitive, I have submitted a corresponding issue with [GCP support](https://cloud.google.com/support-hub). (Actually used PIT)

## Bug Description

When creating a new `BigQueryDataset` resource corresponding to an already existing BigQuery dataset, and omitting the `spec.access` field, seems that config-connector fetches the current set of access rules and updates the `spec.access` field in the cluster, and freezes this list in time.

This is a very surprising behavior, firstly because the `spec.access` field gets populated server-side (I would expect something like `status.access` to be populated instead) which causes a conflict if the user eventually introduces a `spec.access` field in a subsequent update, and secondly because lots of conflicts will be caused through other processes that change the dataset access list from other sources, like `Organization`/`Folder` level policy changes, or access grants through the Pantheon/Cloud Console BigQuery views/GSuite integrations/other APIs.

I would expect not setting `spec.access` to leave the access for a BigQuery dataset in an unmanaged state and changes to the access list from other sources should not cause any conflicts.

Related: #127

### Kubernetes Cluster Version

```
Server Version: v1.18.16-gke.2100
```

### Config Connector Version

```
1.49.1
```

### Steps to reproduce the issue

1. Identify an existing `BigQueryDataset` that was created without KCC, and has a non-empty access list.
2. Create a YAML file matching the existing `BigQueryDataset`
3. Try changing the access list through some place like Pantheon/Cloud Console

### YAML snippets

(Not really applicable, any `BigQueryDataset` would cause this issue)

**Comment by toumorokoshi** [comment]
> When creating a new BigQueryDataset resource corresponding to an already existing BigQuery dataset, and omitting the spec.access field, seems that config-connector fetches the current set of access rules and updates the spec.access field in the cluster, and freezes this list in time.

Can you clarify "freezes this list in time"? It sounds like you're saying:

- `spec.access` cannot be updated by a subsequent `kubectl apply`
- in the case that `spec.access` is unmanaged, the field does not update with subsequent updates to DataSet access direclty against the GCP API.

Are either of those accurate descriptions?

> This is a very surprising behavior, firstly because the spec.access field gets populated server-side (I would expect something like status.access to be populated instead) which causes a conflict if the user eventually introduces a spec.access field in a subsequent update, and secondly because lots of conflicts will be caused through other processes that change the dataset access list from other sources, like Organization/Folder level policy changes, or access grants through the Pantheon/Cloud Console BigQuery views/GSuite integrations/other APIs.

Can you explain the nature of the conflict? I would imagine that the behavior for setting the `spec.access` field and then applying would overwrite the value that was retrieved from the server. 

+1 to status being a better candidate for writing out values were retrieved from the server. But IIUC, this is a general KCC idiom to backfill the spec field.

We could eventually pursue changing this behavior in a v2, but I don't believe it's a practical fix for the near future.

**Comment by dflemstr** [comment]
Hey! I submitted some additional details in the internal Google buganizer ticket, I will try to keep the ticket and this issue in better sync :)

> Can you clarify "freezes this list in time"? It sounds like you're saying:
> 
>     * `spec.access` cannot be updated by a subsequent `kubectl apply`

Exactly, my observation so far (haven't exhaustively tried all the corner cases though) has been that the `spec.access` field gets populated server-side, both by existing access lists (in the case when the KRM object "acquires" an existing cloud resource), as well as any folder/org level policies that get inherited by the BigQueryDataset.

Any subsequent `kubectl apply` that adds a `spec.access` will not be reconcilable with the cluster `spec.access` if the cases mentioned above cause some sort of conflict (e.g. when the applied `spec.access` isn't compliant with the folder/org level policy).

>     * in the case that `spec.access` is unmanaged, the field does not update with subsequent updates to DataSet access direclty against the GCP API.

Yes, exactly.  Described from another perspective: changes through the upstream GCP API are reverted by a subsequent reconciliation by KCC.  This would in and of itself not be a problem for resources where `spec.access` is populated by the application developer, but I think it is surprising behavior when `spec.access` is not populated.

> Can you explain the nature of the conflict? I would imagine that the behavior for setting the `spec.access` field and then applying would overwrite the value that was retrieved from the server.

I wonder how this would work together with inherited permissions from parent resources? 

If `spec.access` is supposed to be "the complete list" of members that have access to the resource, then would the correct behavior be to either have the applied list of members override any inherited access, i.e. if I apply access with member A, but there is an inherited access for member B, the effective `spec.access` would contain only A.  In this case, what happens when the parent resource policy changes, e.g. to add inherited access for member C?

I see one possible behavior to be that `spec.access` only contains the members that are specific to that resource, and doesn't include inherited members.  Hence, the behavior would be for `spec.access` to be purely additive and for the list to be merged with whatever upstream list comes from the GCP API.

> 
> +1 to status being a better candidate for writing out values were retrieved from the server. But IIUC, this is a general KCC idiom to backfill the spec field.
> 
> We could eventually pursue changing this behavior in a v2, but I don't believe it's a practical fix for the near future.

Agreed that what you describe would be a breaking change, and I believe we can find another short-term solution that maintains the current API, for example by taking the additive approach mentioned above.

**Comment by jcanseco** [comment]
Hi @lostick, we have managed to provide a solution to the original issue, but your use-case (being able to manage only a subset of the list) is unfortunately going to be more difficult to support. Let us know if such a feature represents a critical blocker for you.

For the original issue, the problem was that users couldn't actually allow list fields to be externally-managed. That is, even if you didn't specify a list field, if it's set on the GCP resource, KCC will reflect it back onto the KCC resource, and then pin the list to that value henceforth. The reasons it works this way are complicated and arise from a combination of technical limitations coming from both K8s and KCC. 

We were able to solve this problem by allowing users to prevent KCC from reflecting back fields from the GCP resource onto the KCC resource entirely if not specified by the user. You can do this by setting the new `cnrm.cloud.google.com/state-into-spec` annotation to `absent` (default value is `merge`, which means to allow unspecified fields to be reflected back).

In other words, we introduced an annotation that allows the user to keep the desired and observed states separate. Indeed, separating these two states is the long-term direction we want to take KCC towards, since we've experienced a lot of problems with the current behavior where the desired and observed states are mixed in the `spec`.

**Comment by lostick** [comment]
Hi @jcanseco, thanks for letting me know about the `cnrm.cloud.google.com/state-into-spec` annotation. I see that this annotation was introduced in `1.58.0`.

From what I understand, when the value is set to `absent`, the spec field does not get updated with the current GCP state. As you said, this only fixes the aforementioned issue but does not allow a merge-in-place without overriding existing values.

I believe we'll have to either make the spec field unmanaged (and use the[ bigquery API](https://pkg.go.dev/cloud.google.com/go/bigquery#hdr-Datasets_and_Tables) directly), or enforce setting accesses from the CRD. Thanks!

**Comment by jcanseco** [comment]
> From what I understand, when the value is set to absent, the spec field does not get updated with the current GCP state.

Yes that is correct.

And no problem. Happy to help :)
